File: services/api/src/main/java/net/blueshell/api/controller/CommitteeController.java
package net.blueshell.api.controller;

import jakarta.validation.Valid;
import jakarta.ws.rs.NotFoundException;
import net.blueshell.api.base.AdvancedController;
import net.blueshell.api.base.DTO;
import net.blueshell.api.common.enums.Role;
import net.blueshell.api.dto.committee.AdvancedCommitteeDTO;
import net.blueshell.api.mapping.committee.AdvancedCommitteeMapper;
import net.blueshell.api.mapping.committee.SimpleCommitteeMapper;
import net.blueshell.api.model.Committee;
import net.blueshell.api.service.CommitteeService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.List;

@RestController
public class CommitteeController extends AdvancedController<CommitteeService, AdvancedCommitteeMapper, SimpleCommitteeMapper> {

    @Autowired
    public CommitteeController(CommitteeService service, AdvancedCommitteeMapper advancedCommitteeMapper, SimpleCommitteeMapper simpleCommitteeMapper) {
        super(service, advancedCommitteeMapper, simpleCommitteeMapper);
    }

    @GetMapping("/committees")
    public List<? extends DTO> getCommittees(@RequestParam(required = false) boolean isMember) {
        if (getPrincipal() != null && getPrincipal().hasRole(Role.BOARD)) {
            return advancedMapper.toDTOs(service.findAll());
        } else if (isMember) {
            return advancedMapper.toDTOs(service.findAllById(new ArrayList<>(getPrincipal().getCommitteeIds())));
        }

        return simpleMapper.toDTOs(service.findAll());
    }

    @PreAuthorize("hasAuthority('BOARD')")
    @PostMapping("/committees")
    public DTO createCommittee(@Valid @RequestBody AdvancedCommitteeDTO advancedCommitteeDTO) {
        Committee committee = advancedMapper.fromDTO(advancedCommitteeDTO);
        System.out.println("committee: " + committee);
        service.createCommittee(committee);
        return advancedMapper.toDTO(committee);
    }

    @PreAuthorize("hasPermission(#committeeId, 'Committee', 'write')")
    @PutMapping(value = "/committees/{committeeId}")
    public DTO updateCommittee(
            @PathVariable("committeeId") Long committeeId,
            @Valid @RequestBody AdvancedCommitteeDTO advancedCommitteeDTO) {
        Committee oldCommittee = service.findById(committeeId);

        if (!oldCommittee.hasMember(getPrincipal()) && !hasAuthority(Role.BOARD)) {
            throw new NotFoundException();
        }

        advancedCommitteeDTO.setId(committeeId);
        Committee newCommittee = advancedMapper.fromDTO(advancedCommitteeDTO);
        System.out.println("newCommittee: " + newCommittee);
        service.update(newCommittee);
        return advancedMapper.toDTO(newCommittee);
    }

    @PreAuthorize("hasPermission(#committeeId, 'Committee', 'delete')")
    @DeleteMapping(value = "/committees/{committeeId}")
    public void deleteCommitteeById(@PathVariable("committeeId") Long committeeId) {
        service.deleteById(committeeId);
    }
}

File: api/src/main/kotlin/net/blueshell/api/committee/web/CommitteeController.kt
package net.blueshell.api.committee.web

import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.annotation.security.PermitAll
import jakarta.validation.Valid
import net.blueshell.api.shared.enums.Role
import net.blueshell.api.shared.web.AdvancedController
import net.blueshell.api.shared.dto.BaseDTO
import net.blueshell.api.committee.web.dto.AdvancedCommitteeDTO
import net.blueshell.api.committee.application.CommitteeService
import net.blueshell.api.committee.persistence.asAdvancedDto
import net.blueshell.api.committee.persistence.asSimpleDto
import net.blueshell.api.committee.web.dto.asEntity
import org.springframework.http.HttpStatus
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@Tag(name = "Committees")
class CommitteeController(
    service: CommitteeService
) : AdvancedController<CommitteeService>(
    service
) {
    @GetMapping("/committeeMembers/committees")
    @PermitAll
    fun findCommitteesForCurrentUser(): MutableList<out BaseDTO> {
        val principalId = principal?.id ?: return mutableListOf()
        if (hasAuthority(Role.BOARD)) {
            return service.findAll().map { it.asAdvancedDto() }.toMutableList()
        }

        val committees = service.findAllByUserId(principalId)
        return committees.map { it.asAdvancedDto() }.toMutableList()
    }

    @GetMapping("/committees")
    @PermitAll
    fun findCommittees(): MutableList<out BaseDTO> {
        if (hasAuthority(Role.BOARD)) {
            return service.findAll().map { it.asAdvancedDto() }.toMutableList()
        }

        return service.findAll().map { it.asSimpleDto() }.toMutableList()
    }

    @PreAuthorize("hasPermission(#committeeId, 'Committee', 'read')")
    @GetMapping("/committees/{committeeId}")
    fun findCommitteeById(
        @PathVariable committeeId: Long
    ): BaseDTO {
        val committee = service.findById(committeeId)
        if (hasAuthority(Role.BOARD) || committee.hasMember(principal)) {
            return committee.asAdvancedDto()
        }

        return committee.asSimpleDto()
    }

    @PreAuthorize("hasAuthority('BOARD')")
    @PostMapping("/committees")
    @ResponseStatus(HttpStatus.CREATED)
    fun createCommittee(@Valid @RequestBody advancedCommitteeDTO: @Valid AdvancedCommitteeDTO): AdvancedCommitteeDTO {
        var committee = advancedCommitteeDTO.asEntity()
        committee = service.create(committee)
        return committee.asAdvancedDto()
    }

    @PreAuthorize("hasAuthority('BOARD') || (#id == #dto.id && hasPermission(#id, 'Committee', 'write'))")
    @PutMapping(value = ["/committees/{id}"])
    fun updateCommittee(
        @PathVariable id: Long,
        @Valid @RequestBody dto: @Valid AdvancedCommitteeDTO
    ): AdvancedCommitteeDTO {
        var committee = service.findById(id)
        dto.asEntity(committee)
        committee = service.update(committee)
        return committee.asAdvancedDto()
    }

    @PreAuthorize("hasAuthority('BOARD')")
    @DeleteMapping(value = ["/committees/{id}"])
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteCommitteeById(@PathVariable id: Long) {
        service.deleteById(id)
    }
}
